File: ios/Classes/request/PowerImageBaseRequest.m
//
//  PowerImageRequest.m
//  power_image
//
//  Created by 王振辉 on 2021/7/14.
//

#import "PowerImageBaseRequest.h"
#import "PowerImageDispatcher.h"
#import "PowerImagePlugin.h"
#import "PowerFlutterMultiFrameImage.h"

@interface PowerImageBaseRequest()

@property (nonatomic, strong) PowerImageResult *result;

@end

@implementation PowerImageBaseRequest

- (instancetype)initWithArguments:(NSDictionary *)arguments {
    self = [super init];
    if (self) {
        _uniqueKey = arguments[@"uniqueKey"];
        _imageRequestConfig = [PowerImageRequestConfig requestConfigWithArguments:arguments];
    }
    return self;
}

- (BOOL)configTask {
    if (!self.imageRequestConfig) {
        self.imageTaskState = PowerImageRequestStateInitializeFailed;
        return NO;
    }
    self.imageTaskState = PowerImageRequestStateInitializeSucceed;
    return YES;
}


- (BOOL)startLoading {
    // TODO 如果开始加载失败  是否要上报状态
    if (![self.imageTaskState isEqualToString:PowerImageRequestStateInitializeSucceed] &&
            ![self.imageTaskState isEqualToString:PowerImageRequestStateLoadFailed]) {
            // 只有初始化好 或者 加载失败的情况可以重新加载
            return NO;
    }
    if (!self.imageRequestConfig) {
       // 保险起见 还是再做一次判空
       return NO;
    }
    
    [self performLoadImage];
    return YES;
}

- (void)performLoadImage {
    __weak typeof(self) weakSelf = self;
    [[PowerImageLoader sharedInstance] handleRequest:self.imageRequestConfig completed:^(PowerImageResult *powerImageResult){
        __strong typeof(self) self = weakSelf;
        [self requestResultWithPowerImageResult:powerImageResult];
    }];
}

- (void)requestResultWithPowerImageResult:(PowerImageResult *)powerImageResult {
    self.result = powerImageResult;
}

- (void)onLoadSuccess {
    __weak typeof(self) weakSelf = self;
    [[PowerImageDispatcher sharedInstance] runOnMainThread:^{
       __strong typeof(self) self = weakSelf;
       self.imageTaskState = PowerImageRequestStateLoadSucceed;
       [[PowerImagePlugin sharedInstance] sendImageStateEvent:[self encode] success:YES];
    }];
}

- (void)onLoadFailed:(NSString *)errMsg {
    __weak typeof(self) weakSelf = self;
    [[PowerImageDispatcher sharedInstance] runOnMainThread:^{
        __strong typeof(self) self = weakSelf;
        self.imageTaskState = PowerImageRequestStateLoadFailed;
        NSMutableDictionary *event = [self encode];
        event[@"errMsg"] = errMsg;
        [[PowerImagePlugin sharedInstance] sendImageStateEvent:event success:NO];
    }];
}

- (BOOL)stopTask {
    return false;
}

- (NSMutableDictionary *)encode {
    NSMutableDictionary *encodedTask = [[NSMutableDictionary alloc] init];
    encodedTask[@"uniqueKey"] = self.uniqueKey;
    encodedTask[@"state"] = self.imageTaskState;
    if (self.result != nil && self.result.success && [self.result.image isKindOfClass:[PowerFlutterMultiFrameImage class]]) {
        encodedTask[@"_multiFrame"] = @YES;
    }
    return encodedTask;
}

@end

File: android/src/main/java/com/taobao/power_image/request/PowerImageBaseRequest.java
package com.taobao.power_image.request;

import com.taobao.power_image.PowerImageEngineContext;
import com.taobao.power_image.dispatcher.PowerImageDispatcher;
import com.taobao.power_image.loader.FlutterMultiFrameImage;
import com.taobao.power_image.loader.PowerImageLoader;
import com.taobao.power_image.loader.PowerImageLoaderProtocol;
import com.taobao.power_image.loader.PowerImageResult;

import java.util.HashMap;
import java.util.Map;

/**
 * created by Muke on 2021/7/21
 */
public abstract class PowerImageBaseRequest {
    protected static final String REQUEST_STATE_INITIALIZE_SUCCEED = "initializeSucceed";
    protected static final String REQUEST_STATE_INITIALIZE_FAILED = "initializeFailed";
    protected static final String REQUEST_STATE_LOAD_SUCCEED = "loadSucceed";
    protected static final String REQUEST_STATE_LOAD_FAILED = "loadFailed";
    protected static final String REQUEST_STATE_RELEASE_SUCCEED = "releaseSucceed";
    protected static final String REQUEST_STATE_RELEASE_FAILED = "releaseFailed";

    public static final String RENDER_TYPE_EXTERNAL = "external";
    public static final String RENDER_TYPE_TEXTURE = "texture";

    private final PowerImageEngineContext engineContext;
    private PowerImageRequestConfig imageRequestConfig;
    String requestId;
    protected String imageTaskState;
    protected PowerImageResult realResult;

    public PowerImageBaseRequest(PowerImageEngineContext context, Map<String, Object> arguments) {
        engineContext = context;
        requestId = (String) arguments.get("uniqueKey");
        imageRequestConfig = PowerImageRequestConfig.requestConfigWithArguments(arguments);
    }

    public boolean configTask() {
        boolean inited = imageRequestConfig != null;
        imageTaskState = inited ? REQUEST_STATE_INITIALIZE_SUCCEED
                : REQUEST_STATE_INITIALIZE_FAILED;
        return inited;
    }

    public boolean startLoading() {
        if (!REQUEST_STATE_INITIALIZE_SUCCEED.equals(imageTaskState)
                && !REQUEST_STATE_LOAD_FAILED.equals(imageTaskState)) {
            // 只有初始化好 或者 加载失败的情况可以重新加载
            return false;
        }
        if (imageRequestConfig == null) {
            return false;
        }
        performLoadImage();
        return true;
    }

    private void performLoadImage() {
        PowerImageLoader.getInstance().handleRequest(
                imageRequestConfig,
                new PowerImageLoaderProtocol.PowerImageResponse() {
                    @Override
                    public void onResult(PowerImageResult result) {
                        PowerImageBaseRequest.this.onLoadResult(result);
                    }
                }
        );
    }

    void onLoadResult(PowerImageResult result) {
        this.realResult = result;
    }

    public void onLoadSuccess() {
        PowerImageDispatcher.getInstance().runOnMainThread(new Runnable() {
            @Override
            public void run() {
                PowerImageBaseRequest.this.imageTaskState = REQUEST_STATE_LOAD_SUCCEED;
                engineContext.getPowerImageEventSink()
                        .sendImageStateEvent(PowerImageBaseRequest.this.encode(), true);
            }
        });
    }

    public void onLoadFailed(final String errMsg) {
        PowerImageDispatcher.getInstance().runOnMainThread(new Runnable() {
            @Override
            public void run() {
                PowerImageBaseRequest.this.imageTaskState = REQUEST_STATE_LOAD_FAILED;
                Map<String, Object> event = PowerImageBaseRequest.this.encode();
                event.put("errMsg", errMsg != null ? errMsg : "failed!");
                engineContext.getPowerImageEventSink()
                        .sendImageStateEvent(event, false);
            }
        });
    }

    public boolean stopTask() {
        return false;
    }

    public Map<String, Object> encode() {
        Map<String, Object> encodedTask = new HashMap<>();
        encodedTask.put("uniqueKey", requestId);
        encodedTask.put("state", imageTaskState);
        if (realResult != null && realResult.success && realResult.image instanceof FlutterMultiFrameImage) {
            encodedTask.put("_multiFrame", true);
        }
        return encodedTask;
    }

}
